org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/bundle/BundleSource.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) OSGi Alliance (2020). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.osgi.test.junit5.bundle;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.junit.jupiter.params.provider.ArgumentsSource;
+
+@Target({
+	ElementType.ANNOTATION_TYPE, ElementType.METHOD
+})
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@ArgumentsSource(BundleArgumentsProvider.class)
+public @interface BundleSource {
+
+	/**
+	 * Optional SymbolicNameFilter used to filter Bundles by SymbolicName using
+	 * regular expression pattern.
+	 *
+	 * @return The symbolicNamePattern Strings.
+	 */
+	String[] symbolicNamePattern() default {};
+
+	/**
+	 * Optional single states used to filter Bundles.
+	 *
+	 * @return The states.
+	 */
+	int[] states() default {};
```

Comment:
Why do you need an array of states when you can use a bitmask? All the bundle state values are discrete bits. See BundleTracker's use of stateMask. https://docs.osgi.org/specification/osgi.core/7.0.0/util.tracker.html#org.osgi.util.tracker.BundleTracker.BundleTracker-BundleContext-int-BundleTrackerCustomizer-

Reply:
A shame, but first time i worked with a bit-mask in java. thank you for the hint.